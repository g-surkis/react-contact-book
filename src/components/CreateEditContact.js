import React, { Component } from "react";
import PropTypes from "prop-types";
import { addContactToLocStorage } from "../services/contacts";
import { validateObj, updateContactsInLocStorage } from "../services/contacts";

import { connect } from "react-redux";
import {
  editContactAction,
  updateStoreAfterAddingContact
} from "../actions/actions";

import avatar from "../static/avatar.jpg";

class CreateEditContact extends Component {
  constructor(props) {
    super(props);
    this.state = {
      firstName: "",
      lastName: "",
      tel: "",
      email: "",
      birthDate: null
    };

    this.onSubmit = this.onSubmit.bind(this);
    this.onInputChange = this.onInputChange.bind(this);
  }
  componentDidMount() {
    if (this.props.label === "Save Contact") {
      let contact = this.props.contact;
      this.setState({
        firstName: contact.firstName,
        lastName: contact.lastName,
        tel: contact.phone,
        email: contact.email,
        birthDate: contact.birthday
      });
    }
  }
  onInputChange(event) {
    event.preventDefault();
    const name = event.target.name;
    const value = event.target.value;
    this.setState({ [name]: value });
  }
  onSubmit() {
    let obj = {
      firstName: this.state.firstName,
      lastName: this.state.lastName,
      phone: this.state.tel,
      email: this.state.email,
      birthday: this.state.birthDate
    };

    if (validateObj(obj)) {
      if (this.props.label === "Save Contact") {
        let contacts = this.props.contacts;
        let newContacts = [...contacts];
        newContacts[this.props.contactKey] = obj;
        this.props.editContacts(newContacts);
        updateContactsInLocStorage(newContacts);
      } else {
        addContactToLocStorage(obj);
        this.props.addContact(obj);
      }
    } else {
     console.log('Enter correct data');
    }
  }

  
  componentWillUnmount() {
    this.onSubmit();
  }

  render() {
    return (
      <div>
        <form onSubmit={this.onSubmit}>
          <p className="form_row">
            <label className="form_label">
              {" "}
              First Name{" "}
              <input
                className="form_input"
                type="text"
                name="firstName"
                defaultValue={
                  this.props.contact !== undefined
                    ? this.props.contact.firstName
                    : ""
                }
                onChange={this.onInputChange}
              />
            </label>
          </p>
          <p className="form_row">
            <label className="form_label">
              {" "}
              Last Name{" "}
              <input
                className="form_input"
                type="text"
                name="lastName"
                defaultValue={
                  this.props.contact !== undefined
                    ? this.props.contact.lastName
                    : ""
                }
                onChange={this.onInputChange}
              />
            </label>
          </p>
          <p className="form_row">
            <label className="form_label">
              {" "}
              Phone Number{" "}
              <input
                className="form_input"
                type="tel"
                name="tel"
                pattern="[+][0-9]{12}"
                defaultValue={
                  this.props.contact !== undefined
                    ? this.props.contact.phone
                    : ""
                }
                onChange={this.onInputChange}
              />
            </label>
          </p>
          <p className="form_row">
            <label className="form_label">
              {" "}
              Email Address{" "}
              <input
                className="form_input"
                type="email"
                name="email"
                defaultValue={
                  this.props.contact !== undefined
                    ? this.props.contact.email
                    : ""
                }
                onChange={this.onInputChange}
              />
            </label>
          </p>
          <p className="form_row">
            <label className="form_label">
              {" "}
              Date of birth{" "}
              <input
                className="form_input"
                type="date"
                name="birthDate"
                defaultValue={
                  this.props.contact !== undefined
                    ? this.props.contact.birthday
                    : ""
                }
                onChange={this.onInputChange}
              />
            </label>
          </p>
        </form>
        <div className="form_img">
          <img src={avatar} alt="avatar"/>
        </div>
        <div className="btn_area">{this.props.button}</div>
      </div>
    );
  }
}

export default connect(
  function mapStateToProps(state) {
    return {
      contacts: state.contactsReducer
    };
  },
  function mapDispatchToProps(dispatch) {
    return {
      addContact: obj => {
        dispatch(updateStoreAfterAddingContact(obj));
      },
      editContacts: obj => {
        dispatch(editContactAction(obj));
      }
    };
  }
)(CreateEditContact);

CreateEditContact.propTypes = {
  addContact: PropTypes.func,
  button: PropTypes.object,
  contact: PropTypes.object,
  contactKey: PropTypes.number,
  editContacts: PropTypes.func,
  label: PropTypes.string,
  trigger: PropTypes.object
};
