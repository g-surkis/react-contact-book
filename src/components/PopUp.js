import React from "react";
import Popup from "reactjs-popup";
import CreateEditContact from "./CreateEditContact";

export default props => (
  <Popup trigger={props.trigger} position="right center" modal>
    {close => (
      <div className="modal">
        <CreateEditContact
          button={
            <button
              type="sabmit"
              className="animated_btns"
              onClick={() => {
                close();
              }}
            >
              {props.label}
            </button>
          }
          contact={props.contact}
          contactKey={props.contactKey}
          {...props}
        />
      </div>
    )}
  </Popup>
);
