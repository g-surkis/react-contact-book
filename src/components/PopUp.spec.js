import React from "react";
import { shallow } from "enzyme";
import CreateEditContact from "./CreateEditContact";

import PopUp from "./PopUp";

describe("PopUp component", () => {
  const mockFetchGetCLose = jest.fn();
  const props = {
    contact: {},
    contactKey: 0,
    label: "",
    close: mockFetchGetCLose
  };
  let popUp = shallow(<PopUp {...props} />);

  
  it("renders properly", () => {
    expect(popUp).toMatchSnapshot();
  });
});
