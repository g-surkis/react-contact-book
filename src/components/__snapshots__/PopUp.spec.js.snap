// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopUp component renders properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <_default
    close={[MockFunction]}
    contact={Object {}}
    contactKey={0}
    label=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "arrow": true,
      "arrowStyle": Object {},
      "children": [Function],
      "className": "",
      "closeOnDocumentClick": true,
      "closeOnEscape": true,
      "contentStyle": Object {},
      "defaultOpen": false,
      "disabled": false,
      "keepTooltipInside": false,
      "lockScroll": false,
      "modal": true,
      "mouseEnterDelay": 100,
      "mouseLeaveDelay": 100,
      "offsetX": 0,
      "offsetY": 0,
      "on": Array [
        "click",
      ],
      "onClose": [Function],
      "onOpen": [Function],
      "open": false,
      "overlayStyle": Object {},
      "position": "right center",
      "trigger": null,
    },
    "ref": null,
    "rendered": [Function],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "arrow": true,
        "arrowStyle": Object {},
        "children": [Function],
        "className": "",
        "closeOnDocumentClick": true,
        "closeOnEscape": true,
        "contentStyle": Object {},
        "defaultOpen": false,
        "disabled": false,
        "keepTooltipInside": false,
        "lockScroll": false,
        "modal": true,
        "mouseEnterDelay": 100,
        "mouseLeaveDelay": 100,
        "offsetX": 0,
        "offsetY": 0,
        "on": Array [
          "click",
        ],
        "onClose": [Function],
        "onOpen": [Function],
        "open": false,
        "overlayStyle": Object {},
        "position": "right center",
        "trigger": null,
      },
      "ref": null,
      "rendered": [Function],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
